data-generation/src/main/scala/thesis/SnapshotDelta.scala

```diff
@@ -46,19 +46,26 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
 
   override def snapshotAtTime(instant: Instant): AttributeGraph = {
 
-    val closestGraph = graphs.reduce(returnClosestGraph(instant))
+    //Closest graph should be based on number of logs, not time.
+    // But for it to be feasible each graph should have a sequence ID as well
+    // This is possible, I think, but not implemented
+    val closestGraph = graphs.reduceLeft((snap1, snap2) =>
+      if (Duration.between(snap2.instant, instant).isNegative) { // This means the next snapshot in the list is past the instant
+        snap1
+      } else {
+        snap2
+      })
     logger.warn(s"Instant $instant, Closest :graph${closestGraph.instant}")
     if (closestGraph.instant == instant) {
       logger.warn("The queried graph is already materialized")
       closestGraph.graph
-    } else if (closestGraph.instant.isBefore(instant)) {
-      logger.warn("Closest graph is in the future. Need to apply logs forwards")
+    } else if (instant < closestGraph.instant) { // We are asking an instant before the initial materialized snapshot
+      logger.warn("Closest graph in the future, and there is no one in the past.")
+      createGraph(logs.map(l => (l.timestamp, l)).filterByRange(Instant.MIN, instant).map(_._2))
```

Comment:
def getLogsBetween(logs: RDD[LogTSV], t1:Instant, t2: Instant): RDD[LogTSV] = {
      logs.map(log => (log.timestamp, log)).filterByRange(t1, t2))
```
Kan vi ekstrahere denne loggfiltreringen til en egen funksjon? Vi gjør vel den flere steder også(?) 
Isolerer den bedre for en eventuell test og

Reply:
nja trekker det tilbake, det er jo veldig deklarativt sånn det er egentlig. man lager basically en eksakt wrapper rundt sparkfunksjonalitet if that makes sense